<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Catherine Benavides
- Ayudante: Nicolás Ojeda, Eduardo Moya

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Luis Montero M.
- Nombre de alumno 2:


### **Link de repositorio de GitHub:** `https://github.com/Luiandmont/Laboratorios_MDS_2024`

### Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 7 días desde la publicación, 3 días de atraso con 1 punto de descuento c/u.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos Y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquer material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

### Librerias utiles para el Lab 😸

In [176]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
#!pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

# 1. Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos.
El objetivo principal será el de realizar un trabajo de limpieza de prepración y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.




In [177]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = 'Dirección donde tiene los archivos en el Drive'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```





In [178]:
df = pd.read_json('students_grades.json')
df.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course math score reading score writing score  
0  standard                    none         37            46            46  
1  standard                    none        124           142           142  
2  standard                    none        156           196           193  
3  standard               completed         80            78            81  
4  standard                    none        131           166           166


<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **875** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **875** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier alumno que posea un valor diferente a esta escala, debería ser eliminado de la base de datos.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?


10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [179]:
def exploratory_data_analysis(dataframe):
    """
    Descripción: 
    -----------
    Función de exploración de las característcas del dataset de trabajo

    Parámetros:
    -----------
    La función recibe un dataframe de pandas
    """
    dataframe.columns = dataframe.columns.str.replace(' ', '_')
    nums = [pd.to_numeric(dataframe[i]) for i in dataframe.columns[-3:]]
    nums = pd.DataFrame(nums).T
    categs = dataframe.iloc[:,:6].astype('category')
    df = pd.concat([categs,nums],axis=1)
    dims = df.shape
    nom_cols = df.columns.to_list()
    prim = df.head(5)
    ulti = df.tail(5)
    muestra = df.sample(5, random_state=42)
    descrip = df.describe().round(2)
    nulos = df.isna().sum()
    unic = df.nunique()
    nduplic = df.duplicated(subset=['names']).sum()
    casos_duplic = df.loc[df.duplicated(subset=['names']),:].sort_values('names', ascending=True)

    print("\nRESUMEN DATAFRAME\n","\n\n1. Dimensiones del dataframe: "f'{dims[0]} filas y {dims[1]} columnas',"\n\n2. Nombre columnas:"f'{nom_cols}')
    print("\n3. Primeros 5 casos:")
    display(prim)
    print("4. Últimos 5 casos :")
    display(ulti)
    print("5. Muestra aleatoria de 5 casos:")
    display(muestra)
    print("6. Resumen descriptivos puntajes pruebas:")
    display(descrip)
    print("\n\n7. Valores nulos por columna:\n"f'{nulos}''\n',"\n8. Valores únicos por columna:\n"f'{unic}''\n',"\n9. Duplicados: "f'{nduplic} casos''\n')
    print("10. Listado de casos duplicados:\n")
    display(casos_duplic)

    pass


In [180]:
exploratory_data_analysis(df)


RESUMEN DATAFRAME
 

1. Dimensiones del dataframe: 1400 filas y 9 columnas 

2. Nombre columnas:['names', 'gender', 'race/ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course', 'math_score', 'reading_score', 'writing_score']

3. Primeros 5 casos:


names  gender race/ethnicity parental_level_of_education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test_preparation_course  math_score  reading_score  writing_score  
0  standard                    none        37.0           46.0           46.0  
1  standard                    none       124.0          142.0          142.0  
2  standard                    none       156.0          196.0          193.0  
3  standard               completed        80.0           78.0           81.0  
4  standard                    none       131.0          166.0          166.0

4. Últimos 5 casos :


names  gender race/ethnicity parental_level_of_education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test_preparation_course  math_score  reading_score  \
1395      standard                    none       193.0          142.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1398  free/reduced                    none       193.0          196.0   
1399      standard               completed        68.0           80.0   

      writing_score  
1395          193.0  
1396           81.0  
1397           62.0  
1398          193.0  
1399           76.0

5. Muestra aleatoria de 5 casos:


names  gender race/ethnicity parental_level_of_education  \
665    Bonnie Williams  female        group A            some high school   
624        Marcia York  female        group D           bachelor's degree   
115  Francisco Johnson    male        group B          associate's degree   
478      Edward Lowman    male        group E          associate's degree   
233     Thomas Simmons    male        group C           bachelor's degree   

            lunch test_preparation_course  math_score  reading_score  \
665      standard               completed       193.0          196.0   
624  free/reduced                    none        63.0           73.0   
115      standard               completed        82.0           84.0   
478  free/reduced               completed        91.0           73.0   
233      standard               completed       131.0          171.0   

     writing_score  
665           97.0  
624           78.0  
115           78.0  
478           80.0  
233          196.0

6. Resumen descriptivos puntajes pruebas:


math_score  reading_score  writing_score
count     1360.00        1360.00        1353.00
mean        89.54          93.09          91.85
std         47.15          46.39          47.04
min          0.00          17.00          10.00
25%         59.00          63.00          62.00
50%         72.00          76.00          75.00
75%         97.25         100.00         100.00
max        196.00         196.00         196.00



7. Valores nulos por columna:
names                           0
gender                          0
race/ethnicity                  0
parental_level_of_education     0
lunch                           0
test_preparation_course         0
math_score                     40
reading_score                  40
writing_score                  47
dtype: int64
 
8. Valores únicos por columna:
names                          1153
gender                            2
race/ethnicity                    5
parental_level_of_education       6
lunch                             2
test_preparation_course           2
math_score                       96
reading_score                    88
writing_score                    92
dtype: int64
 
9. Duplicados: 247 casos

10. Listado de casos duplicados:



names  gender race/ethnicity parental_level_of_education  \
441       Aaron Oakden    male        group D                 high school   
1393       Adam Romero    male        group C                 high school   
1395  Alexander Dillon    male        group D          associate's degree   
1049     Amanda Barker  female        group C                some college   
1174         Amy Hawes  female        group B           bachelor's degree   
...                ...     ...            ...                         ...   
768   William Gardiner    male        group E                 high school   
1121        William Ng    male        group C            some high school   
1023     William Saska    male        group B          associate's degree   
476        Willie Neff  female        group E          associate's degree   
1355     Zachary Glenn    male        group E                some college   

             lunch test_preparation_course  math_score  reading_score  \
441   free/reduced                    none         NaN          193.0   
1393      standard               completed       124.0          193.0   
1395      standard                    none       193.0          142.0   
1049  free/reduced               completed        64.0           85.0   
1174      standard                    none       142.0          147.0   
...            ...                     ...         ...            ...   
768   free/reduced                    none        55.0           56.0   
1121      standard                    none        62.0           64.0   
1023      standard                    none       193.0           42.0   
476   free/reduced                    none         NaN          193.0   
1355      standard               completed        99.0           87.0   

      writing_score  
441           124.0  
1393           49.0  
1395          193.0  
1049           85.0  
1174          193.0  
...             ...  
768            51.0  
1121           55.0  
1023          142.0  
476           142.0  
1355           81.0  

[247 rows x 9 columns]

In [181]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df.melt(
        value_vars=["math_score", "reading_score", "writing_score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

**Justificación:**

```
Escriba aquí su justificación
```
**R.** Se observan datos anómalos (posiblemente errores de registro) en scores en matemática, lectura y escritura, los que en cada caso, son alrededor del 25% de los registros. Eso es lo que revelan los boxplots, ya que el tercer cuartil (o percentil 75) de la distribución de las puntuaciones en las variables está en los 100 puntos, que es el máximo de la escala de evaluación. 

**Respuesta Esperada:**

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y cree una función que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**

- [ ] Crear función que permita limpiar el dataset.
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hint**

Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.


**Respuesta Esperada:**

In [ ]:
def clean_data(dataframe):
    dataframe.columns = dataframe.columns.str.replace(' ', '_')
    nums = [pd.to_numeric(dataframe[i]) for i in dataframe.columns[-3:]]
    nums = pd.DataFrame(nums).T
    categs = dataframe.iloc[:,:6].astype('category')
    df = pd.concat([categs,nums],axis=1)
    df = df.drop_duplicates(subset=['names'], keep='first')
    df = df.loc[(df.loc[:, "math_score"] <= 100) & (df.loc[:, "reading_score"] <= 100) & (df.loc[:, "writing_score"] <= 100)]
    
    return df.reset_index(drop=True)

In [ ]:
df_c= clean_data(df)

In [ ]:
df_c

names  gender race/ethnicity parental_level_of_education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test_preparation_course  math_score  reading_score  \
0        standard                    none        37.0           46.0   
1        standard               completed        80.0           78.0   
2        standard                    none        73.0           66.0   
3        standard               completed        81.0           75.0   
4        standard               completed        94.0           86.0   
..            ...                     ...         ...            ...   
870      standard                    none        69.0           75.0   
871  free/reduced               completed        23.0           44.0   
872  free/reduced               completed        66.0           74.0   
873      standard               completed        54.0           61.0   
874      standard               completed        68.0           80.0   

     writing_score  
0             46.0  
1             81.0  
2             63.0  
3             76.0  
4             87.0  
..             ...  
870           71.0  
871           36.0  
872           81.0  
873           62.0  
874           76.0  

[875 rows x 9 columns]

> Utilice la función creada en el punto 1.1 para verificar sus resultados.

In [ ]:
exploratory_data_analysis(df_c)


RESUMEN DATAFRAME
 

1. Dimensiones del dataframe: 875 filas y 9 columnas 

2. Nombre columnas:['names', 'gender', 'race/ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course', 'math_score', 'reading_score', 'writing_score']

3. Primeros 5 casos:


names  gender race/ethnicity parental_level_of_education  \
0     Rita Courtney  female        group B            some high school   
1  Charles Linstrom    male        group A           bachelor's degree   
2       Brian Young    male        group C            some high school   
3    Howard Jimenez    male        group E            some high school   
4      Wayne Wilson    male        group B            some high school   

      lunch test_preparation_course  math_score  reading_score  writing_score  
0  standard                    none        37.0           46.0           46.0  
1  standard               completed        80.0           78.0           81.0  
2  standard                    none        73.0           66.0           63.0  
3  standard               completed        81.0           75.0           76.0  
4  standard               completed        94.0           86.0           87.0

4. Últimos 5 casos :


names  gender race/ethnicity parental_level_of_education  \
870   Richard Young    male        group D                 high school   
871   Wanda Russell  female        group B                 high school   
872  Marina Zeigler  female        group C           bachelor's degree   
873   Laurie Carter  female        group B            some high school   
874    Amanda Perez  female        group A                 high school   

            lunch test_preparation_course  math_score  reading_score  \
870      standard                    none        69.0           75.0   
871  free/reduced               completed        23.0           44.0   
872  free/reduced               completed        66.0           74.0   
873      standard               completed        54.0           61.0   
874      standard               completed        68.0           80.0   

     writing_score  
870           71.0  
871           36.0  
872           81.0  
873           62.0  
874           76.0

5. Muestra aleatoria de 5 casos:


names  gender race/ethnicity parental_level_of_education  \
797      Donna Duarte  female        group C                 high school   
319  Edward Hutcheson    male        group A             master's degree   
382    Louis Mckinzie    male        group C          associate's degree   
790     Allen Schwarz    male        group C            some high school   
422       Kara Greene  female        group D            some high school   

            lunch test_preparation_course  math_score  reading_score  \
797      standard                    none        65.0           69.0   
319  free/reduced                    none        73.0           74.0   
382      standard                    none        74.0           73.0   
790      standard               completed        63.0           60.0   
422      standard               completed        61.0           74.0   

     writing_score  
797           67.0  
319           72.0  
382           67.0  
790           57.0  
422           72.0

6. Resumen descriptivos puntajes pruebas:


math_score  reading_score  writing_score
count      875.00         875.00         875.00
mean        66.23          69.20          68.11
std         15.08          14.63          15.17
min          0.00          17.00          10.00
25%         57.00          59.00          57.50
50%         66.00          70.00          69.00
75%         76.50          79.00          79.00
max        100.00         100.00         100.00



7. Valores nulos por columna:
names                          0
gender                         0
race/ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64
 
8. Valores únicos por columna:
names                          875
gender                           2
race/ethnicity                   5
parental_level_of_education      6
lunch                            2
test_preparation_course          2
math_score                      79
reading_score                   72
writing_score                   77
dtype: int64
 
9. Duplicados: 0 casos

10. Listado de casos duplicados:



Empty DataFrame
Columns: [names, gender, race/ethnicity, parental_level_of_education, lunch, test_preparation_course, math_score, reading_score, writing_score]
Index: []

**Justificación:**

```
RESPUESTA:

Se observa una limpieza completa para los datos contenidos en el dataset, pues para los scores de matemática, lectura y escritura, ningún puntaje supera los 100 que es el límite superior del rango en que se mueven las puntuaciones. Por otra parte, la suma de duplicados es cero, y el dataframe que contiene dichos casos (casos duplicados), está vacío (cero registro). 

```

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [ ]:
def nota_chilena(dataframe_in, columnas):
  dataframe_in.loc[:,columnas] = dataframe_in.loc[:,columnas].apply(lambda x: x*0.06 +1).round(2)

  return dataframe_in

In [ ]:
df_grades=nota_chilena(df_c, ['math_score','reading_score','writing_score'])
df_grades

names  gender race/ethnicity parental_level_of_education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test_preparation_course  math_score  reading_score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing_score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
870           5.26  
871           3.16  
872           5.86  
873           4.72  
874           5.56  

[875 rows x 9 columns]

**Respuesta esperada:**

In [ ]:
df_grades

names  gender race/ethnicity parental_level_of_education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test_preparation_course  math_score  reading_score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing_score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
870           5.26  
871           3.16  
872           5.86  
873           4.72  
874           5.56  

[875 rows x 9 columns]

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [ ]:
def alumnos_promedio(dataframe_in):
  dataframe_in['GPA'] = dataframe_in[['math_score','reading_score','writing_score']].mean(axis=1).round(2)

  return dataframe_in

In [ ]:
df_grades = alumnos_promedio(df_grades)

In [ ]:
df_grades.head()

names  gender race/ethnicity parental_level_of_education  \
0     Rita Courtney  female        group B            some high school   
1  Charles Linstrom    male        group A           bachelor's degree   
2       Brian Young    male        group C            some high school   
3    Howard Jimenez    male        group E            some high school   
4      Wayne Wilson    male        group B            some high school   

      lunch test_preparation_course  math_score  reading_score  writing_score  \
0  standard                    none        3.22           3.76           3.76   
1  standard               completed        5.80           5.68           5.86   
2  standard                    none        5.38           4.96           4.78   
3  standard               completed        5.86           5.50           5.56   
4  standard               completed        6.64           6.16           6.22   

    GPA  
0  3.58  
1  5.78  
2  5.04  
3  5.64  
4  6.34

In [ ]:
df_grades['GPA'].describe().round(2)

count    875.00
mean       5.07
std        0.85
min        1.54
25%        4.50
50%        5.12
75%        5.66
max        7.00
Name: GPA, dtype: float64

**Respuesta Esperada:**

In [ ]:
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental_level_of_education  \
0     Rita Courtney  female        group B            some high school   
1  Charles Linstrom    male        group A           bachelor's degree   
2       Brian Young    male        group C            some high school   
3    Howard Jimenez    male        group E            some high school   
4      Wayne Wilson    male        group B            some high school   

      lunch test_preparation_course  math_score  reading_score  writing_score  \
0  standard                    none        3.22           3.76           3.76   
1  standard               completed        5.80           5.68           5.86   
2  standard                    none        5.38           4.96           4.78   
3  standard               completed        5.86           5.50           5.56   
4  standard               completed        6.64           6.16           6.22   

    GPA  
0  3.58  
1  5.78  
2  5.04  
3  5.64  
4  6.34

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta:**

In [ ]:
def reprobados_destacados(dataframe_in):
    df_copy = dataframe_in.copy()
    mascara = df_copy.loc[:,'GPA'] < 4.00
    reprobados = df_copy.loc[mascara,:]
    df_copy.loc[:,'GPA'] = np.where(df_copy.loc[:,'GPA'] < 4.00, 'R', df_copy.loc[:,'GPA'])
    aprobados = df_copy.loc[~mascara,:].sort_values('GPA', ascending=False)
    
    return reprobados, df_copy, aprobados


In [ ]:
df1, df2, df3 = reprobados_destacados(df_grades)

In [ ]:
df1

names  gender race/ethnicity parental_level_of_education  \
0      Rita Courtney  female        group B            some high school   
7        Mikki Amaya  female        group A                 high school   
16     Samuel Dougal    male        group C             master's degree   
39   Rodger Talavera    male        group B                 high school   
43    Morris Johnson    male        group A          associate's degree   
..               ...     ...            ...                         ...   
852   Charles Harian    male        group C            some high school   
854    Wendy Gunnell  female        group D            some high school   
868      Diana Costa  female        group C                 high school   
869      Amy Sieving  female        group C                 high school   
871    Wanda Russell  female        group B                 high school   

            lunch test_preparation_course  math_score  reading_score  \
0        standard                    none        3.22           3.76   
7    free/reduced               completed        3.04           3.88   
16   free/reduced               completed        3.76           3.52   
39   free/reduced                    none        3.16           2.74   
43   free/reduced               completed        3.40           4.30   
..            ...                     ...         ...            ...   
852  free/reduced               completed        3.70           4.12   
854  free/reduced                    none        2.62           3.04   
868  free/reduced                    none        3.46           3.76   
869      standard                    none        2.74           2.74   
871  free/reduced               completed        2.38           3.64   

     writing_score   GPA  
0             3.76  3.58  
7             3.46  3.46  
16            3.76  3.68  
39            2.62  2.84  
43            4.18  3.96  
..             ...   ...  
852           3.94  3.92  
854           2.92  2.86  
868           3.58  3.60  
869           2.80  2.76  
871           3.16  3.06  

[86 rows x 10 columns]

In [ ]:
df2

names  gender race/ethnicity parental_level_of_education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test_preparation_course  math_score  reading_score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing_score   GPA  
0             3.76     R  
1             5.86  5.78  
2             4.78  5.04  
3             5.56  5.64  
4             6.22  6.34  
..             ...   ...  
870           5.26   5.3  
871           3.16     R  
872           5.86  5.42  
873           4.72  4.54  
874           5.56  5.48  

[875 rows x 10 columns]

In [ ]:
df3

names  gender race/ethnicity parental_level_of_education  \
435       Richard Pryor    male        group E           bachelor's degree   
838       Sandra Pompey  female        group E          associate's degree   
253   Elizabeth Beasley  female        group E           bachelor's degree   
313           Emma Gray  female        group E           bachelor's degree   
795      Angela Cressey  female        group D                some college   
..                  ...     ...            ...                         ...   
76          John Salter    male        group D                some college   
770     Carrie Robinson  female        group C                 high school   
226  Florencia Morrison  female        group A          associate's degree   
710      Shirley Ybanez  female        group C            some high school   
199        Jim Burkhart    male        group D           bachelor's degree   

            lunch test_preparation_course  math_score  reading_score  \
435      standard               completed        7.00           7.00   
838      standard                    none        7.00           7.00   
253      standard                    none        7.00           7.00   
313      standard               completed        6.94           7.00   
795      standard                    none        6.88           7.00   
..            ...                     ...         ...            ...   
76       standard                    none        3.64           4.24   
770  free/reduced                    none        3.10           4.66   
226  free/reduced                    none        3.22           4.42   
710      standard               completed        3.64           4.06   
199      standard                    none        4.24           3.94   

     writing_score   GPA  
435           7.00   7.0  
838           7.00   7.0  
253           7.00   7.0  
313           7.00  6.98  
795           6.94  6.94  
..             ...   ...  
76            4.18  4.02  
770           4.24   4.0  
226           4.36   4.0  
710           4.30   4.0  
199           3.82   4.0  

[789 rows x 10 columns]

In [ ]:
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)
dataframe1

names  gender race/ethnicity parental_level_of_education  \
0      Rita Courtney  female        group B            some high school   
7        Mikki Amaya  female        group A                 high school   
16     Samuel Dougal    male        group C             master's degree   
39   Rodger Talavera    male        group B                 high school   
43    Morris Johnson    male        group A          associate's degree   
..               ...     ...            ...                         ...   
852   Charles Harian    male        group C            some high school   
854    Wendy Gunnell  female        group D            some high school   
868      Diana Costa  female        group C                 high school   
869      Amy Sieving  female        group C                 high school   
871    Wanda Russell  female        group B                 high school   

            lunch test_preparation_course  math_score  reading_score  \
0        standard                    none        3.22           3.76   
7    free/reduced               completed        3.04           3.88   
16   free/reduced               completed        3.76           3.52   
39   free/reduced                    none        3.16           2.74   
43   free/reduced               completed        3.40           4.30   
..            ...                     ...         ...            ...   
852  free/reduced               completed        3.70           4.12   
854  free/reduced                    none        2.62           3.04   
868  free/reduced                    none        3.46           3.76   
869      standard                    none        2.74           2.74   
871  free/reduced               completed        2.38           3.64   

     writing_score   GPA  
0             3.76  3.58  
7             3.46  3.46  
16            3.76  3.68  
39            2.62  2.84  
43            4.18  3.96  
..             ...   ...  
852           3.94  3.92  
854           2.92  2.86  
868           3.58  3.60  
869           2.80  2.76  
871           3.16  3.06  

[86 rows x 10 columns]

In [ ]:
dataframe2

names  gender race/ethnicity parental_level_of_education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test_preparation_course  math_score  reading_score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing_score   GPA  
0             3.76     R  
1             5.86  5.78  
2             4.78  5.04  
3             5.56  5.64  
4             6.22  6.34  
..             ...   ...  
870           5.26   5.3  
871           3.16     R  
872           5.86  5.42  
873           4.72  4.54  
874           5.56  5.48  

[875 rows x 10 columns]

In [ ]:
dataframe3

names  gender race/ethnicity parental_level_of_education  \
435       Richard Pryor    male        group E           bachelor's degree   
838       Sandra Pompey  female        group E          associate's degree   
253   Elizabeth Beasley  female        group E           bachelor's degree   
313           Emma Gray  female        group E           bachelor's degree   
795      Angela Cressey  female        group D                some college   
..                  ...     ...            ...                         ...   
76          John Salter    male        group D                some college   
770     Carrie Robinson  female        group C                 high school   
226  Florencia Morrison  female        group A          associate's degree   
710      Shirley Ybanez  female        group C            some high school   
199        Jim Burkhart    male        group D           bachelor's degree   

            lunch test_preparation_course  math_score  reading_score  \
435      standard               completed        7.00           7.00   
838      standard                    none        7.00           7.00   
253      standard                    none        7.00           7.00   
313      standard               completed        6.94           7.00   
795      standard                    none        6.88           7.00   
..            ...                     ...         ...            ...   
76       standard                    none        3.64           4.24   
770  free/reduced                    none        3.10           4.66   
226  free/reduced                    none        3.22           4.42   
710      standard               completed        3.64           4.06   
199      standard                    none        4.24           3.94   

     writing_score   GPA  
435           7.00   7.0  
838           7.00   7.0  
253           7.00   7.0  
313           7.00  6.98  
795           6.94  6.94  
..             ...   ...  
76            4.18  4.02  
770           4.24   4.0  
226           4.36   4.0  
710           4.30   4.0  
199           3.82   4.0  

[789 rows x 10 columns]

**Respuesta:**

```
Escriba aqui su respuesta
```

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [ ]:
# Código para quintiles

df_grades['GPA_cuantiles'] = pd.qcut(
    df_grades["GPA"],
    q=5,
    labels=['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']) 

df_grades

names  gender race/ethnicity parental_level_of_education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test_preparation_course  math_score  reading_score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing_score   GPA GPA_cuantiles  
0             3.76  3.58      Bajísimo  
1             5.86  5.78          Alto  
2             4.78  5.04         Medio  
3             5.56  5.64          Alto  
4             6.22  6.34      Muy alto  
..             ...   ...           ...  
870           5.26  5.30         Medio  
871           3.16  3.06      Bajísimo  
872           5.86  5.42          Alto  
873           4.72  4.54          Bajo  
874           5.56  5.48          Alto  

[875 rows x 11 columns]

In [ ]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_grades.sort_values(by="GPA"),
    x="GPA",
    y="parental_level_of_education",
    hover_data=["gender"],
    color="GPA_cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig


```
Escriba aqui su respuesta
```
**R.** La respuesta no puede ser del todo concluyente sobre la relación entre nivel educacional de los padres y el desempeño académicos de los estudiantes, dado que las notas para los quintiles de rendimiento de los estudiantes por nivel educacional de los padres, son similares. Sin embargo, sí se observan patrones de comportamiento diferenciado según la primera de las variables. A medida que el nivel educacional de los padres es más alto, se observa menor dispersión en el rendimiento académico: Mientras en los niveles educativos más bajos (some high school) se observa, en términos relativos, mayor dispersión del rendimiento, en los niveles más altos (bachelor y master), el rendimiento muestra menores grados de dispersión. Y otro patrón que se puede observar en los datos, se relaciona con la dispersión en el rendimiento en los extremos de la escala: En el nivel Bajísimo y Muy alto, se registra más dispersión que la observada en los niveles intermedios de la misma, sobre todo en los niveles educacionales más bajos.    

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

In [ ]:
# Inserte su respuesta aquí

dfs1 = pd.read_json('students_grades_1.json')
dfs2 = pd.read_json('students_grades_2.json')


In [ ]:
def merge_dataframes(dataframe1, dataframe2, axis=0):
    df = pd.concat([dataframe1, dataframe2], axis=0).reset_index(drop=True)
    return df

In [ ]:
dff = merge_dataframes(dfs1, dfs2)
dff

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
870           5.26  
871           3.16  
872           5.86  
873           4.72  
874           5.56  

[875 rows x 9 columns]

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos] **


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


In [ ]:
# SE RELLENAN ESPACIOS EN LOS NOMBRES DE LAS COLUMNAS PARA EVITAR ERRORES O CAÍDAS DEL CÓDIGO.

dff.columns = dff.columns.str.replace(' ', '_')
dff.columns

Index(['names', 'gender', 'race/ethnicity', 'parental_level_of_education',
       'lunch', 'test_preparation_course', 'math_score', 'reading_score',
       'writing_score'],
      dtype='object')

In [ ]:
# CALCULO PROMEDIO DE NOTAS GPA

dff['GPA'] = dff[['math_score','reading_score','writing_score']].mean(axis=1).round(2)

In [ ]:
# SIMPLIFICACIÓN NIVEL EDUCACIONAL PADRES MEDIANTE MAPEO Y FUNCIÓN DE CONVERSIÓN.

def simplify_education_level(level):
    conversion_dict = {
        'some high school': 'school',
        'some college': 'school',
        'high school': 'school',
        "bachelor's degree": 'college',
        "associate's degree": 'college',
        "master's degree": 'postgraduate'
    }
    return conversion_dict.get(level, level)


dff['simple_parental_level_of_education'] = list(map(simplify_education_level, dff['parental_level_of_education']))


In [ ]:
t1 = dff.groupby(['race/ethnicity','simple_parental_level_of_education']).agg({'math_score':'mean','reading_score':'mean','writing_score':'mean','GPA':'mean'}).round(2).reset_index()

In [ ]:
t2 = dff.groupby(['race/ethnicity','simple_parental_level_of_education']).agg(count=('simple_parental_level_of_education','count'), percentage=('simple_parental_level_of_education', lambda x: (x.count()/dff.shape[0])*100)).round(2).reset_index()

In [ ]:
tf = pd.merge(
    left=t1, 
    right=t2,
    how='left'
)
tf

race/ethnicity simple_parental_level_of_education  math_score  \
0         group A                            college        4.74   
1         group A                       postgraduate        4.69   
2         group A                             school        4.57   
3         group B                            college        5.07   
4         group B                       postgraduate        4.91   
5         group B                             school        4.69   
6         group C                            college        5.02   
7         group C                       postgraduate        4.92   
8         group C                             school        4.76   
9         group D                            college        5.11   
10        group D                       postgraduate        5.22   
11        group D                             school        5.02   
12        group E                            college        5.54   
13        group E                       postgraduate        5.54   
14        group E                             school        5.40   

    reading_score  writing_score   GPA  count  percentage  
0            5.00           4.89  4.88     24        2.74  
1            5.23           5.35  5.09      2        0.23  
2            4.73           4.56  4.62     51        5.83  
3            5.26           5.19  5.18     54        6.17  
4            5.69           5.55  5.38      5        0.57  
5            4.89           4.76  4.78    107       12.23  
6            5.37           5.35  5.25    102       11.66  
7            5.14           5.10  5.06     15        1.71  
8            5.02           4.92  4.90    155       17.71  
9            5.25           5.25  5.20     70        8.00  
10           5.54           5.73  5.50     20        2.29  
11           5.13           5.11  5.09    149       17.03  
12           5.45           5.45  5.48     52        5.94  
13           6.03           5.89  5.82      6        0.69  
14           5.31           5.16  5.29     63        7.20

In [ ]:
fig = px.box(
    dff,
    x="simple_parental_level_of_education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
)
fig.show()

**ANÁLISIS RESULTADOS GRÁFICO**  

El análisis del gráfico muestra que según el nivel educacional de los padres, se observan diferencias en el rendimiento de los estudiantes según su grupo étnico. Si se toma como referencia la nota 5, se puede apreciar que en el caso del nivel "school", en 2 grupos étnicos (B y A) la mediana de rendimiento no supera la nota 5, y en el grupo étnico C, la mediana de rendimiento es igual a 5. Para el caso de los padres con nivel educacional "college" se observa un mejor rendimiento en todos los grupos étnicos: Todas las medianas de rendimiento están por sobre la nota 5. La excepción en este nivel es el grupo étnico A, cuya mediana de rendimiento para este nivel educativo de los padres, sigue bajo la nota 5. Para el nivel educacional "postgraduate", se observa que los grupos B y E presentan los mejores rendimientos (tienen las medianas más altas), aunque con un alto nivel de variación: el rango intercuartil, representado por la altura de las cajas, en este nivel, aumenta de manera considerable. Ahora, se debe tener en cuenta que la alta variabilidad observada se relaciona con la baja cantidad de estudiantes con padres que cuentan con este nivel educativo. También se puede anotar que recién a este nivel educacional, el grupo A de estudiantes muestra una mejora en el rendimiento académico, pues la mediana de GPA supera la nota 5 de referencia. Se aprecia como patrón que si bien hay un efecto del nivel educacional de los padres en el rendimiento académico de los estudiantes, al interior de estos grupos también se generan diferencias según grupo étnico de pertenencia. En todos los niveles, tiende a mostrar mejor rendimiento relativo, el grupo B, y peor rendimiento, el grupo A.

En síntesis, a nivel descriptivo, se observa que nivel educacional de los padres afecta el rendimiento académico de los estudiantes, y que parte de la variación al interior de los grupos formados por los niveles educacionales de los padres, esta asociada al grupo étnico al que pertenecen las familias de los estudiantes. 


**Resultado Esperado de la tabla**

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.74 |            5    |            4.89 |  4.88 |      24 | 2.74 %       |
|  1 |           | postgraduate                         |         4.69 |            5.23 |            5.35 |  5.09 |       2 | 0.23 %       |
|  2 |           | school                               |         4.57 |            4.73 |            4.56 |  4.62 |      51 | 5.83 %       |
|  3 | group B          | college                              |         5.07 |            5.26 |            5.19 |  5.18 |      54 | 6.17 %       |
|  4 |           | postgraduate                         |         4.91 |            5.69 |            5.55 |  5.38 |       5 | 0.57 %       |
|  5 |           | school                               |         4.69 |            4.89 |            4.76 |  4.78 |     107 | 12.23 %      |
|  6 | group C          | college                              |         5.02 |            5.37 |            5.35 |  5.25 |     102 | 11.66 %      |
|  7 |           | postgraduate                         |         4.92 |            5.14 |            5.1  |  5.06 |      15 | 1.71 %       |
|  8 |           | school                               |         4.76 |            5.02 |            4.92 |  4.9  |     155 | 17.71 %      |
|  9 | group D          | college                              |         5.11 |            5.25 |            5.25 |  5.2  |      70 | 8.0 %        |
| 10 |           | postgraduate                         |         5.22 |            5.54 |            5.73 |  5.5  |      20 | 2.29 %       |
| 11 |           | school                               |         5.02 |            5.13 |            5.11 |  5.09 |     149 | 17.03 %      |
| 12 | group E          | college                              |         5.54 |            5.45 |            5.45 |  5.48 |      52 | 5.94 %       |
| 13 |          | postgraduate                         |         5.54 |            6.03 |            5.89 |  5.82 |       6 | 0.69 %       |
| 14 |          | school                               |         5.4  |            5.31 |            5.16 |  5.29 |      63 | 7.2 %        |

### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene nuevamente los alumnos de los registros corruptos que ud. y su equipo filtraron en el análisis anterior.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

**RESPUESTA**  

Para el caso de evaluación sobre cuál es el mejor join a utilizar para generar el nuevo dataframe, según los resultados arrojados por cada tipo de join, se opta por el left join. Esto porque permite conservar de mejor forma la integralidad de los datos. Es decir, en el proceso de unión de las tablas, no perdemos ningún caso del dataframe df_grades  original (que se trabajó, limpió y transformó con tanto trabajo y sentido analítico), y se gana todos los datos de las nuevas columnas al momento de usar como llave los nombres de los estudiantes (aunque si en el segundo dataframe no hay concidencias en base a la llave utilizada, se generarán datos NAN). Para el caso particular de este trabajo, el inner join arroja los mismos resultados que el left join, pues en ambas tablas hay coincidencias en base a la llave utilizada, razón por la cual se conserva la integralidad de la tabla df_grades. Por las razones opuestas, se descarta el right y el outer join: se generan casos NAN en los atributos, perdiendose integralidad de los datos.

In [ ]:
# Inserte su respuesta aquí

dfo = pd.read_csv('other_grades.csv')
dfo

names  science score  history score
0           Pam Otoole             57            4.0
1          Diane Olson             48            4.8
2       Sarah Flannery             72            6.1
3     Stanley Chappell             47            3.7
4        Arnold Dodson             60            5.2
..                 ...            ...            ...
995    Melanie Acevedo             79            6.3
996   Christopher Wall             62            4.0
997  Margaret Spannaus             47            4.0
998     Vickie Sellers             47            5.7
999      Kari Mitchell             53            6.2

[1000 rows x 3 columns]

In [ ]:
# OUTER JOIN

df_out = pd.merge(
    left=df_grades,
    right=dfo,
    on="names",
    how="outer",
)

df_out

names  gender race/ethnicity parental_level_of_education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
..                 ...     ...            ...                         ...   
995       Leona Markow     NaN            NaN                         NaN   
996  Maxine Gulbranson     NaN            NaN                         NaN   
997        Jeane Budds     NaN            NaN                         NaN   
998        Nancy Ramos     NaN            NaN                         NaN   
999     Andrea Mcdavid     NaN            NaN                         NaN   

        lunch test_preparation_course  math_score  reading_score  \
0    standard                    none        3.22           3.76   
1    standard               completed        5.80           5.68   
2    standard                    none        5.38           4.96   
3    standard               completed        5.86           5.50   
4    standard               completed        6.64           6.16   
..        ...                     ...         ...            ...   
995       NaN                     NaN         NaN            NaN   
996       NaN                     NaN         NaN            NaN   
997       NaN                     NaN         NaN            NaN   
998       NaN                     NaN         NaN            NaN   
999       NaN                     NaN         NaN            NaN   

     writing_score   GPA GPA_cuantiles  science score  history score  
0             3.76  3.58      Bajísimo             25            3.8  
1             5.86  5.78          Alto             56            5.2  
2             4.78  5.04         Medio             58            3.8  
3             5.56  5.64          Alto             72            5.0  
4             6.22  6.34      Muy alto             94            4.6  
..             ...   ...           ...            ...            ...  
995            NaN   NaN           NaN             39            3.8  
996            NaN   NaN           NaN             38            3.9  
997            NaN   NaN           NaN             43            3.9  
998            NaN   NaN           NaN             54            6.2  
999            NaN   NaN           NaN             40            4.2  

[1000 rows x 13 columns]

In [ ]:
# LEFT JOIN

df_left = pd.merge(
    left=df_grades,
    right=dfo,
    on="names",
    how="left",
)

df_left

names  gender race/ethnicity parental_level_of_education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test_preparation_course  math_score  reading_score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing_score   GPA GPA_cuantiles  science score  history score  
0             3.76  3.58      Bajísimo             25            3.8  
1             5.86  5.78          Alto             56            5.2  
2             4.78  5.04         Medio             58            3.8  
3             5.56  5.64          Alto             72            5.0  
4             6.22  6.34      Muy alto             94            4.6  
..             ...   ...           ...            ...            ...  
870           5.26  5.30         Medio             62            4.2  
871           3.16  3.06      Bajísimo             16            3.6  
872           5.86  5.42          Alto             59            4.1  
873           4.72  4.54          Bajo             43            4.3  
874           5.56  5.48          Alto             61            5.3  

[875 rows x 13 columns]

In [ ]:
# RIGHT JOIN

df_rg = pd.merge(
    left=df_grades,
    right=dfo,
    on="names",
    how="right",
)

df_rg

names  gender race/ethnicity parental_level_of_education  \
0           Pam Otoole  female        group B           bachelor's degree   
1          Diane Olson  female        group C                some college   
2       Sarah Flannery  female        group B             master's degree   
3     Stanley Chappell    male        group A          associate's degree   
4        Arnold Dodson    male        group C                some college   
..                 ...     ...            ...                         ...   
995    Melanie Acevedo  female        group E             master's degree   
996   Christopher Wall    male        group C                 high school   
997  Margaret Spannaus  female        group C                 high school   
998     Vickie Sellers  female        group D                some college   
999      Kari Mitchell  female        group D                some college   

            lunch test_preparation_course  math_score  reading_score  \
0        standard                    none        5.32           5.32   
1        standard               completed        5.14           6.40   
2        standard                    none        6.40           6.70   
3    free/reduced                    none        3.82           4.42   
4        standard                    none        5.56           5.68   
..            ...                     ...         ...            ...   
995      standard               completed        6.28           6.94   
996  free/reduced                    none        4.72           4.30   
997  free/reduced               completed        4.54           5.26   
998      standard               completed        5.08           5.68   
999  free/reduced                    none        5.62           6.16   

     writing_score   GPA GPA_cuantiles  science score  history score  
0             5.44  5.36          Alto             57            4.0  
1             6.28  5.94      Muy alto             48            4.8  
2             6.58  6.56      Muy alto             72            6.1  
3             3.64  3.96      Bajísimo             47            3.7  
4             5.50  5.58          Alto             60            5.2  
..             ...   ...           ...            ...            ...  
995           6.70  6.64      Muy alto             79            6.3  
996           4.30  4.44          Bajo             62            4.0  
997           4.90  4.90          Bajo             47            4.0  
998           5.62  5.46          Alto             47            5.7  
999           6.16  5.98      Muy alto             53            6.2  

[1000 rows x 13 columns]

In [ ]:
# INNER JOIN

df_in = pd.merge(
    left=df_grades,
    right=dfo,
    on="names",
    how="inner",
)

df_in

names  gender race/ethnicity parental_level_of_education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test_preparation_course  math_score  reading_score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing_score   GPA GPA_cuantiles  science score  history score  
0             3.76  3.58      Bajísimo             25            3.8  
1             5.86  5.78          Alto             56            5.2  
2             4.78  5.04         Medio             58            3.8  
3             5.56  5.64          Alto             72            5.0  
4             6.22  6.34      Muy alto             94            4.6  
..             ...   ...           ...            ...            ...  
870           5.26  5.30         Medio             62            4.2  
871           3.16  3.06      Bajísimo             16            3.6  
872           5.86  5.42          Alto             59            4.1  
873           4.72  4.54          Bajo             43            4.3  
874           5.56  5.48          Alto             61            5.3  

[875 rows x 13 columns]

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>